Update REST Sensors to allow json_attributes_template for state attributes (MQTT style)

shbatm opened the pull request:
## Description:

1) Add `json_attributes` and  `force_update` options to the `rest` platform's `binary_sensor` platform (already exist in `sensor`)
2) Update the `rest` `sensor` and `binary_sensor` devices to allow MQTT-style `json_attribute_template` options to break out the JSON response into state attributes.  Currently only first-level JSON dict values can be handled as device state attributes, this allows you to add a nested JSON response object as state values and then (coupled with `json_attributes`) filter *that* object further. [MQTT Documentation](https://www.home-assistant.io/components/sensor.mqtt/#json-attributes-template-configuration) for reference. 

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#9722

## Example entry for `configuration.yaml` (if applicable):
```yaml
binary_sensor:
  - platform: rest
    scan_interval: 600
    resource: !secret resource_url
    name: VPN
    device_class: connectivity
    value_template: "{{ value_json['user']['connected'] }}"
    headers:
      User-Agent: Home Assistant
      Content-Type: application/json
    json_attributes:
      bytes_read: "{{ value_json.connection.bytes_read }}"
      bytes_write: "{{ value_json.connection.bytes_write }}"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [N/A] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [N/A] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [N/A] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

ronmichel commented:
Why can't the actual `json_attributes` be a list of templates? That would make the sensor much more flexible (since the desired attributes might not be nested in the same object).

shbatm commented:
@ronmichel you could in theory, but I was trying to stay within the established architecture by mimicking what the MQTT sensor has already.

balloob commented:
We should not have Jinja2 templates be used to generate JSON values.

pvizeli commented:
@balloob it's what MQTT also does, but yeah it's strange. We should handle that for that kind of integration same over the place.

shbatm commented:
Would it be better to just allow a list of JSON paths (including any nested objects) in the config and return those directly?

Like @pvizeli said, I was just extending something already used in MQTT which works and had a similar need for me with REST.

Do you want me to open an architecture issue for this if you're wanting to be consistent across the board? (I can't think of any other 'generic' integrations that might need this besides MQTT, REST, and maybe command line?).

balloob commented:
We have complex template rendering. So we could do something like this:

```yaml
attributes:
  something: {{ json_value.bla }}
  hello: {{ json_value.yo }} 
```

shbatm commented:
I updated the code to allow the use of complex templates, but I don't see a method for complex templates "with possible json values" -- I have to convert the JSON to a dict and pass that to the render_complex function.  It works, however, you end up with `"{{ connection.bytes_read }}"` instead of `"{{ value_json.connection.bytes_read }}"`.  Am I missing that function somewhere?

shbatm commented:
Is it possible to update [`homeassistant.helpers.template` Line 69](https://github.com/home-assistant/home-assistant/blob/797196dce97d1418e1217fbe645e7da1c94ce6c5/homeassistant/helpers/template.py#L69) from:

```python
    return value.async_render(variables)
```
to:
```python
    return value.async_render_with_possible_json_value(variables)
```
Or duplicate this function to `render_complex_with_possible_json_value`?

MartinHjelmare commented:
Please rebase on latest dev branch.

shbatm committed:
Update Rest Sensors to use json_attributes_template (MQTT style)


Remove debug statement.


Update tests for rest sensors to include json_attributes_template


Update to correct new tests and correct error type to pass tox


Use complex templates.


Work-around for no JSON complex template rendering.

shbatm committed:
Formatting with Black

shbatm committed:
Fix test calls.

shbatm committed:
Fix tests, add back removed warning, black formatting

Removed test "test_update_with_json_attrs_not_dict" for sensor since a complex template can be a list or a dict.

MartinHjelmare commented:
@shbatm please run `black` from the project root.
```
black --fast homeassistant tests
```
@balloob can you continue the review?

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -44,6 +47,7 @@
             [HTTP_BASIC_AUTHENTICATION, HTTP_DIGEST_AUTHENTICATION]
         ),
         vol.Optional(CONF_HEADERS): {cv.string: cv.string},
+        vol.Optional(CONF_JSON_ATTRS): vol.Any(cv.template_complex, cv.template),
```
We should call it `attributes` and it should accept a `{str: cv.template_complex}`

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -52,6 +56,7 @@
         vol.Optional(CONF_USERNAME): cv.string,
         vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
         vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(CONF_FORCE_UPDATE, default=DEFAULT_FORCE_UPDATE): cv.boolean,
```
this seems unrelated to this PR

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -142,3 +168,42 @@ def is_on(self):
+            if value:
+                try:
+                    if isinstance(self._json_attrs, template.Template):
+                        attr = self._json_attrs.render_with_possible_json_value(value)
```
Since a lot of rendering is happening here, it should not happen inside a sync context. We should make this method `aync def async_update(self)` and it should run the rest update in an executor:

```python
await self.hass.async_add_executor_job(self.rest.update)
```

balloob replied:
Also, remove the fact that it can just be a template. `attr` should only be allowed to be a dictionary.

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -142,3 +168,42 @@ def is_on(self):
         self.rest.update()
+        value = self.rest.data
+
+        if self._json_attrs:
```
Turn this code around with the value renderer

balloob replied:
(it should also not be called `json_attrs` as it has nothing to do with that)

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -142,3 +168,42 @@ def is_on(self):
+                    )
+            else:
+                _LOGGER.warning("Empty reply found when expecting JSON data")
+        if value is not None and self._value_template is not None:
```
Adding this here means you need to remove it from the `is_on` property.

balloob commented on homeassistant/components/rest/binary_sensor.py:
```diff
@@ -142,3 +168,42 @@ def is_on(self):
+                            )
+                            _LOGGER.debug("Erroneous JSON: %s", value)
+                        else:
+                            attr.update(
```
Why would this be an update? `attr` is just an empty dict

balloob commented on homeassistant/components/rest/sensor.py:
```diff
@@ -56,6 +56,7 @@
         vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
         vol.Optional(CONF_USERNAME): cv.string,
         vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+        vol.Optional(CONF_JSON_ATTRS): vol.Any(cv.template_complex, cv.template),
```
Please let's update 1 platform at a time. you can update sensor in a future PR.

frenck commented:
Months have passed since last update on this PR, so it is running stale. I'm going to close the PR for now.

Feel free to re-open the PR when you are ready to work on it again 👍
